### ISSUE TITLE
GS-002 — DoS through stale connections

### ISSUE THREAD

**Issue by pospeselr** [blueprint-freespeech/gosling#46]
### Description:
The Gosling protocol does not close stale connections, resulting in resource exhaustion.

### Technical description:
The Gosling protocol uses a polling scheme to update the state of incoming connections to the identity server. For every connection, the stream is polled for data and the state of the handshake is determined.

The current implementation does not drop stale connections, i.e. streams that are asked for data, but do not send any for a long time.

This behavior is implemented in honk_rpc.rs::Session::read_message() on lines 505 and 542:
```Rust
Err(err) => if err.kind() == ErrorKind::WouldBlock || err.kind() == ErrorKind::TimedOut {
   Ok(None)
} else {
   bail!(err);
}
```

If no data is presented, this arm returns Ok(None) and the polling loop continues.

This behavior could be abused by an adversary creating a large number of connections to the identity server. Not only would this result in CPU overhead, since all these connections need to be covered in the loop, but if there are more than ~64k connections, no new connections can be made.

If the adversary drops the connection, the server detects this correctly and the connection is no longer polled.

### Impact:

An adversary could prevent legitimate users from connecting and obtaining an endpoint service_id .

### Recommendation:
• Implement state tracking logic that kills the connection if no bytes are present on the stream after reading the
BSON length header; A BSON message should be sent atomically.
• Enforce handshake completion within a reasonable (short) amount of time, or kill the connection.

**Comment by pospeselr** [comment]
The Honk RPC spec ( https://github.com/blueprint-freespeech/gosling/blob/main/docs/rpc.md ) mentions support for connection timeouts, but they are not currently implemented. As a result, an identity or endpoint server can be easily DoS'd since old connections are never dropped.

**Comment by pospeselr** [comment]
Initial fix in 8c3dfe34d3683df107271ad677c9fcfc61a9e5dc. Future updates will make timeout (and max message size) configurable by embedder.
